experimental/fast_lossless/fast_lossless.cc

```diff
@@ -0,0 +1,665 @@
+
+  if (chunk_size != 16) {
+    auto mask =
+        _mm256_loadu_si256((const __m256i*)(kLaneMask + (16 - chunk_size)));
+    nbits = _mm256_and_si256(nbits, mask);
+    bits = _mm256_and_si256(bits, mask);
+  }
+
+  // Merge nbits and bits from 16-bit to 32-bit lanes.
+  auto nbits_hi16 = _mm256_srli_epi32(nbits, 16);
+  auto nbits_lo16 = _mm256_and_si256(nbits, _mm256_set1_epi32(0xFFFF));
+  auto bits_hi16 = _mm256_srli_epi32(bits, 16);
+  auto bits_lo16 = _mm256_and_si256(bits, _mm256_set1_epi32(0xFFFF));
+
+  nbits = _mm256_add_epi32(nbits_hi16, nbits_lo16);
+  bits = _mm256_or_si256(_mm256_sllv_epi32(bits_hi16, nbits_lo16), bits_lo16);
+
+  // Merge 32 -> 64 bit lanes.
+  auto nbits_hi32 = _mm256_srli_epi64(nbits, 32);
+  auto nbits_lo32 = _mm256_and_si256(nbits, _mm256_set1_epi64x(0xFFFFFFFF));
+  auto bits_hi32 = _mm256_srli_epi64(bits, 32);
+  auto bits_lo32 = _mm256_and_si256(bits, _mm256_set1_epi64x(0xFFFFFFFF));
+
+  nbits = _mm256_add_epi64(nbits_hi32, nbits_lo32);
+  bits = _mm256_or_si256(_mm256_sllv_epi64(bits_hi32, nbits_lo32), bits_lo32);
+
+  alignas(32) uint64_t nbits_simd[4] = {};
+  alignas(32) uint64_t bits_simd[4] = {};
+
+  _mm256_store_si256((__m256i*)nbits_simd, nbits);
+  _mm256_store_si256((__m256i*)bits_simd, bits);
+
+  // Manually merge the buffer bits with the SIMD bits.
+  // Necessary because Write() is only guaranteed to work with <=56 bits.
+  // Trying to SIMD-fy this code results in slower speed (and definitely less
+  // clarity).
+  {
+    for (size_t i = 0; i < 4; i++) {
+      output.buffer |= bits_simd[i] << output.bits_in_buffer;
+      memcpy(output.data.get() + output.bytes_written, &output.buffer, 8);
+      // If >> 64, next_buffer is unused.
+      uint64_t next_buffer = bits_simd[i] >> (64 - output.bits_in_buffer);
+      output.bits_in_buffer += nbits_simd[i];
+      // This `if` seems to be faster than using ternaries.
+      if (output.bits_in_buffer >= 64) {
+        output.buffer = next_buffer;
+        output.bits_in_buffer -= 64;
+        output.bytes_written += 8;
+      }
+    }
+    memcpy(output.data.get() + output.bytes_written, &output.buffer, 8);
+    size_t bytes_in_buffer = output.bits_in_buffer / 8;
+    output.bits_in_buffer -= bytes_in_buffer * 8;
+    output.buffer >>= bytes_in_buffer * 8;
+    output.bytes_written += bytes_in_buffer;
+  }
+}
+#endif
+
+constexpr uint16_t PackSigned(int16_t value) {
+  return (static_cast<uint16_t>(value) << 1) ^
+         ((static_cast<uint16_t>(~value) >> 15) - 1);
+}
+
+struct ChannelRowEncoder {
+  inline void ProcessChunk(const int16_t* row, const int16_t* row_left,
+                           const int16_t* row_top, const int16_t* row_topleft,
+                           size_t chunk_size, BitWriter& output) {
+    if (chunk_size == 0) return;
+    bool continue_rle = true;
+    alignas(32) uint16_t residuals[16] = {};
+    for (size_t ix = 0; ix < chunk_size; ix++) {
+      int16_t px = row[ix];
+      int16_t left = row_left[ix];
```

Comment:
Might be better to use row[ix-1] here, no?
Same with topleft, which should always be the same as top[ic -1].
Maybe the compiler generates more efficient code if it knows that than if the values could come from 4 different places...

Reply:
Ah, but it's not, for row 0 ;)
Either way the emitted code is pretty nice as is so I wouldn't bother.

Reply:
Ah right, for row 0 topleft is a bit weird.
Left is still always current-1 though.
But probably doesn't matter.